src/zir_sema.zig

```diff
@@ -1204,34 +1204,116 @@ fn optionalPayload(
+        }
+        return mod.constInst(scope, unwrap.base.src, .{
+            .ty = operand.ty.castTag(.error_union).?.data.payload,
+            .val = val,
+        });
+    }
+    const b = try mod.requireRuntimeBlock(scope, unwrap.base.src);
+    if (safety_check and mod.wantSafety(scope)) {
+        const is_non_err = try mod.addUnOp(b, unwrap.base.src, Type.initTag(.bool), .is_err, operand);
+        try mod.addSafetyCheck(b, is_non_err, .unwrap_errunion);
+    }
+    return mod.addUnOp(b, unwrap.base.src, operand.ty.castTag(.error_union).?.data.payload, .unwrap_errunion_payload, operand);
 }
 
 /// Pointer in, pointer out
 fn errorUnionPayloadPtr(mod: *Module, scope: *Scope, unwrap: *zir.Inst.UnOp, safety_check: bool) InnerError!*Inst {
     const tracy = trace(@src());
     defer tracy.end();
-    return mod.fail(scope, unwrap.base.src, "TODO implement zir_sema.errorUnionPayloadPtr", .{});
+
+    const operand = try resolveInst(mod, scope, unwrap.positionals.operand);
+    assert(operand.ty.zigTypeTag() == .Pointer);
+
+    if (operand.ty.elemType().zigTypeTag() != .ErrorUnion)
+        return mod.fail(scope, unwrap.base.src, "expected error union type, found {}", .{operand.ty.elemType()});
+
+    const operand_pointer_ty = try mod.simplePtrType(scope, unwrap.base.src, operand.ty.elemType().castTag(.error_union).?.data.payload, !operand.ty.isConstPtr(), .One);
+
+    if (operand.value()) |pointer_val| {
+        const val = try pointer_val.pointerDeref(scope.arena());
+        if (val.getError()) |name| {
+            return mod.fail(scope, unwrap.base.src, "caught unexpected error '{s}'", .{name});
+        }
+        // The same Value represents the pointer to the error union and the payload.
+        return mod.constInst(scope, unwrap.base.src, .{
+            .ty = operand_pointer_ty,
+            .val = pointer_val,
+        });
+    }
+
+    const b = try mod.requireRuntimeBlock(scope, unwrap.base.src);
+    if (safety_check and mod.wantSafety(scope)) {
+        const is_non_err = try mod.addUnOp(b, unwrap.base.src, Type.initTag(.bool), .is_err, operand);
+        try mod.addSafetyCheck(b, is_non_err, .unwrap_errunion);
+    }
+    return mod.addUnOp(b, unwrap.base.src, operand_pointer_ty, .unwrap_errunion_payload_ptr, operand);
 }
 
 /// Value in, value out
 fn errorUnionCode(mod: *Module, scope: *Scope, unwrap: *zir.Inst.UnOp) InnerError!*Inst {
     const tracy = trace(@src());
     defer tracy.end();
-    return mod.fail(scope, unwrap.base.src, "TODO implement zir_sema.errorUnionCode", .{});
+
+    const operand = try resolveInst(mod, scope, unwrap.positionals.operand);
+    if (operand.ty.zigTypeTag() != .ErrorUnion)
+        return mod.fail(scope, unwrap.base.src, "expected error union type, found '{}'", .{operand.ty});
+
+    if (operand.value()) |val| {
+        assert(val.getError() != null);
+        const data = val.castTag(.error_union).?.data;
```

Comment:
This fails with `attempt to use null value` because you never actually wrap a value in this `error_union` value, either you need to make `wrapErrorUnion` actually wrap the value or remove it entirely.
```zig
const a: anyerror!u32 = error.Bar;
@compileLog(a catch |err| @compileLog(@TypeOf(err)));
```

Reply:
Ah I see, I didn't that extra cast.

Reply:
Ok, I think I fixed it by making wrapErrorUnion actually wrap it into an error_union SubValue.